### ISSUE TITLE
Support `az login --identity` for Azure Arc

### ISSUE THREAD

**Issue by jimdigriz** [Azure/azure-cli#16573]
**Is your feature request related to a problem? Please describe.**
For enrolled Azure Arc systems, `azure-cli` does not support pulling the local [managed identity](https://docs.microsoft.com/en-us/azure/azure-arc/servers/managed-identity-authentication).

**Describe the solution you'd like**
`az login --identity` [tests if running on an Azure instance](https://github.com/microsoft/PowerShell-DSC-for-Linux/blob/8f2bca268da6913b4fcbf70ce7f53d8ea86e33a8/Providers/nxOMSAutomationWorker/automationworker/worker/linuxutil.py#L89) and if not to [fallback to using the localhost challenge response endpoint `http://localhost:40342` provided by `/opt/azcmagent/bin/himds`](https://docs.microsoft.com/en-us/azure/azure-arc/servers/managed-identity-authentication#acquiring-an-access-token-using-rest-api).

If azure-cli gets this functionality it means I can have my scripts call `az ...` directly and not have to treat Azure instances differently to on-premise kit; for example to access a shared keyvault or a workspace for logging and metrics.

**Describe alternatives you've considered**
I have to do the REST dance myself and spoof some credentials for the azure-cli via `~/.azure/{accessTokens,azureProfile}.json`.

**Comment by yonzhan** [comment]
az login

**Comment by richeney** [comment]
@jimdigriz We definitely need this. 

In the meantime, do you have any bash code that generates a valid token object in ~/.azure/accessTokens.json using the REST API?

**Comment by jiasli** [comment]
[Using a managed identity with Arc enabled servers](https://docs.microsoft.com/en-us/azure/azure-arc/servers/security-overview#using-a-managed-identity-with-arc-enabled-servers) is supported by Azure Identity in https://github.com/Azure/azure-sdk-for-python/pull/15020. The commit https://github.com/Azure/azure-sdk-for-python/commit/d32f4b3a78d41d4f90e2694c69d62b009e5b5dd4 is released in `azure-identity_1.6.0b2`, `azure-identity_1.6.0b1`, `azure-identity_1.5.0`, but the latest [Azure CLI beta](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-beta) still uses 

https://github.com/Azure/azure-cli/blob/aa456a28c56c8436be1af6e9f2a3e96a21d293f0/src/azure-cli-core/setup.py#L54

which doesn't have this feature.

We will release a new version of Azure CLI beta using `azure-identity_1.6.0b2` with support for managed identity with Arc enabled servers.

**Comment by jiasli** [comment]
⚠ Manipulating `~/.azure/accessTokens.json` is NEVER a supported scenario. `~/.azure/accessTokens.json` will be deprecated in the beta MSAL-based Azure CLI.

> I have to do the REST dance myself and spoof some credentials for the azure-cli via `~/.azure/{accessTokens,azureProfile}.json`.

This approach will stop working immediately once you start using MSAL-based Azure CLI.

**Comment by jimdigriz** [comment]
> @jimdigriz We definitely need this.
> 
> In the meantime, do you have any bash code that generates a valid token object in ~/.azure/accessTokens.json using the REST API?

@richeney Sorry no, I explored this and found the juice was not worth the squeeze; it quickly became apparently that tinkering with `accessTokens.json` was going to be fragile at best.

In the end I baked the logic into my application and moved on.

Maybe you could slum it with shell wrapper for cURL that then goes and calls `az rest --skip-authorization-header --headers authorization=... ...`? Awful but it might help you until MS release a new version of the CLI which could be months...

**Comment by annerajb** [comment]
Any hope of seeing this anytime soon?

Just stumble upon it and spent a while trying to understand why the environment variables where not overriding the 169.254* address. (until it occur to me to check on github)

**Comment by jiasli** [comment]
Due to a recent reorganization, we have to delay the development of Azure Identity and MSAL integration. The rough ETA is the end of this year (2021-12-31), but it is not guaranteed. 

Thanks for understanding.

**Comment by annerajb** [comment]
> Due to a recent reorganization, we have to delay the development of Azure Identity and MSAL integration. The rough ETA is the end of this year (2021-12-31), but it is not guaranteed.
> 
> Thanks for understanding.

is this something that could be sped up by merge request / contributions from the community?

**Comment by yonzhan** [comment]
We plan to hire more people to contribute to this big project:)

**Comment by johlindr** [comment]
Quite strange to leave out the support for MSI on CLI for hybrid scenarios when one of the core benefits on Arc is that the resource gets an Managed Identity...

**Comment by richeney** [comment]
@johlindr, here are some Bash commands for Linux that may help you. Needs jq installed and sudoers configured.

```bash
imds=$(curl -sSL -H Metadata:true http://localhost:40342/metadata/instance?api-version=2020-06-01)
subscriptionId=$(jq -r .compute.subscriptionId <<< $imds)
resourceGroupName=$(jq -r .compute.resourceGroupName <<< $imds)

# Challenge token
challengeTokenPath=$(curl -s -D - -H Metadata:true "http://127.0.0.1:40342/metadata/identity/oauth2/token?api-version=2019-11-01&resource=https%3A%2F%2Fmanagement.azure.com" | grep Www-Authenticate | cut -d "=" -f 2 | tr -d "[:cntrl:]")
challengeToken=$(sudo cat $challengeTokenPath)

# Resource token
token=$(curl -s -H Metadata:true -H "Authorization: Basic $challengeToken" "http://127.0.0.1:40342/metadata/identity/oauth2/token?api-version=2019-11-01&resource=https%3A%2F%2Fmanagement.azure.com" | jq -r .access_token)

# Example ARM REST API call
curl -sSL -X GET -H "Authorization: Bearer $token" -H "Content-Type: application/json" https://management.azure.com/subscriptions/$subscriptionId/resourceGroups/$resourceGroupName/resources?api-version=2020-06-01 | jq .
```

That last command will only work if the Managed Identity has an RBAC role assignment. Here is the command to get the ObjectId for the Managed Identity.

```bash
objectId=$(az connectedmachine show --name <vmname>--resource-group <rgname> --query identity.principalId --output tsv)
```

You can change the resource from management.azure.com to other resources / audiences to get tokens for other REST API endpoints, e.g. `https://vault.azure.net`.

**Comment by johlindr** [comment]
but am I able to use the CLI library functions using  the access token manually retrieved for the MSI?  We are using the "manual" approach aleady to perform certain tasks but need to onboard an Azure Arc Kubernetes cluster and not sure I wish to do that with direct API calls.

**Comment by richeney** [comment]
I'm not going to say anything that isn't already said above:

* No, I don't think you can login with the azure CLI on Arc-enabled servers as the managed identity
* the work to enable this properly is major as it fundamentally changes the auth method - aiming for end of the CY
* others (i.e. @jimdigriz) have tried munging accessTokens.json but have backed away from that as it is fragile

REST API calls it is!

**Comment by jiasli** [comment]
@richeney's comment is correct. This is currently the only way to call Azure APIs from Azure Arc servers.

You may also specify the access token in

```
az rest --headers authorization="<access_token>"
```

While we will implement `az login --identity` for Azure Arc, maybe the "bring your own access token" proposal (https://github.com/Azure/azure-cli/issues/16459) will also make Azure CLI more versatile, so that you may call normal Azure CLI commands with the access token acquired using @richeney's method (https://github.com/Azure/azure-cli/issues/16573#issuecomment-870503359). 🤔

**Comment by richeney** [comment]
@jsntcy / @yonzhan 

Will this be resolved by the July 2021 milestone? 

If so, how will the fix be implemented?

**Comment by jsntcy** [comment]
accidentally close it, reopen now

**Comment by eoq** [comment]
Any update on this?

**Comment by jiasli** [comment]
Nope. We'll make sure to update this thread when there is. In the meantime, might I recommend the **Subscribe** button?

![image](https://user-images.githubusercontent.com/4003950/155913610-6a788ba6-4bbe-4377-b1b4-efa48bc2e680.png)

That way you'll be notified of any updates to this thread, without needlessly pinging everyone on this thread. :)

**Comment by svadimsh** [comment]
any update on that feature?

**Comment by jiasli** [comment]
Current technical difficulties:

- Azure CLI dropped Azure Identity and now directly uses MSAL for user/service principal authentication
- Currently, Azure CLI uses `msrestazure` for managed identity authentication, but `msrestazure` has been replaced by Azure Identity and out of maintainance
- MSAL has no managed identity support, and it is not planed in the near future (https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/58)

In order to support Azure Arc's variation of managed identity, Azure CLI needs to incorporate `msrestazure`'s code and make changed on top of that.

**Comment by jiasli** [comment]
Provided more information in https://github.com/Azure/azure-cli/issues/24150

**Comment by jiasli** [comment]
Azure Arc's managed identity will be supported by MSAL (https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/480) and Azure CLI will migrate its managed identity authentication to MSAL (https://github.com/Azure/azure-cli/pull/25959). After these tasks are done, Azure CLI will support Azure Arc's managed identity, but currently there is no ETA.

**Comment by jjindrich** [comment]
Please, are there some updates ? I need authorize to azure from arc server.

**Comment by bniranjanbhat** [comment]
Hi 
This is a fundamental ask, which has not been worked on yet.
This scenario is supported by powershell AZ modules.
Can you please proiritize this ?

**Comment by msalwen** [comment]
I came across this long-standing and still unresolved issue while trying to figure out how to execute ray {up|down} from an Azure Arc server as ray seems to depend on the azure cli for this. I don't think anything here can help, but if anyone is just trying to access Azure resources from Python code, the following may be useful. It essentially runs the commands that @richeney listed [here](https://github.com/Azure/azure-cli/issues/16573#issuecomment-870503359) but wraps them in Python code and provides a "credential" that we can use to instantiate storage, secrets, and compute management clients. We also use this credential in a slightly different manner to create a pyodbc connection. Here's the code to create the credential and a helper to obtain it that can use the REST API if on ARC or DefaultAzureCredential if not (controlled by an environment variable):
```
#!/usr/bin/env python
"Authenticate against Azure resources with user, managed identity, or Arc server permissions."

import json
import os
import requests

from azure.core.credentials import AccessToken
from azure.identity import DefaultAzureCredential


def _get_access_token(scope):
    """Get access token from a Managed Identities for Azure Resources service.

    Args:
        resource_type (str): Desired Azure scope to access. "storage.azure.com",
            "management.azure.com", "vault.azure.net",
            or "database.windows.net" are common options.

    Returns (azure.core.credentials.AccessToken): Azure access token.
    """

    # Note: The port will need to be parameterized if the service runs elsewhere.
    resp = requests.get(
        "http://127.0.0.1:40342/metadata/identity/oauth2/token?api-version=2019-11-01&"
        f"resource={scope}",
        headers={"Metadata": "true"},
    )
    fn = resp.headers["Www-Authenticate"].split("=")[-1]
    with open(fn) as f:
        token = f.read()
    token = requests.get(
        "http://127.0.0.1:40342/metadata/identity/oauth2/token?api-version=2019-11-01&"
        f"resource={scope}",
        headers={"Metadata": "true", "Authorization": f"Basic {token}"},
    ).content

    token = json.loads(token.decode("UTF-8").strip("\n"))

    return AccessToken(token["access_token"], expires_on=float(token["expires_on"]))


class ARCServerCredential:
    def __init__(self, scope=None):
        """Return credential to work with Arc server.

        Args:
            resource_type (str): Desired Azure scope to access. "storage.azure.com",
                "management.azure.com", "vault.azure.net",
                or "database.windows.net" are common options.
        """
        if scope is None:
            raise ValueError("scope cannot be None.")
        self.access_token = _get_access_token(scope=scope)

    def get_token(self, *scopes, **kwargs):
        return self.access_token


def get_azure_credential(scope=None, **kwargs):
    """Produce a credential accepted by tooling in the Azure SDK for Python.

    Enables authentication using tokens acquired from DefaultAzureCredential
    or, on an Azure Arc server, from an active Managed Identities
    for Azure Resources service.

    Args:
        scope (str): Desired Azure scope to access. Passed to ARCServerCredential.
        kwargs: Arguments to be passed to DefaultAzureCredential.

    Returns (azure.identity.DefaultAzureCredential|ARCServerCredential): Azure access
        token.
    """
    if os.getenv("AZURE_ARC_ENABLED", "false") == "false":
        return DefaultAzureCredential(**kwargs)

    return ARCServerCredential(scope=scope)
```

We have succeeded in using this approach within containers running on the Arc server. This only requires a couple of flags when executing `docker run`: `--network=host --mount type=bind,source=/var/opt/azcmagent,destination=/var/opt/azcmagent,readonly`

If you haven't baked the required env variable into the image, you'll also need to add `--env AZURE_ARC_ENABLED=true` to the run command

**Comment by jiasli** [comment]
@msalwen, if you are retrieving the access token on Azure Arc by yourself, it is recommended to use MSAL: https://msal-python.readthedocs.io/en/latest/#managed-identity

**Comment by richeney** [comment]
Now that [Azure CLI 2.30.0 has switched to MSAL](https://learn.microsoft.com/cli/azure/msal-based-azure-cli), and [MSAL supports Azure Arc](https://learn.microsoft.com/entra/msal/dotnet/advanced/managed-identity), will this finally get past the deadlock?

@jiasli, will the preview of #29187 enable `az login --identity`?

**Comment by husamhilal** [comment]
Any updates on ETA? This is a very critical feature to have on Arc-enabled servers. Thank you very much!

**Comment by msalwen** [comment]
> [@msalwen](https://github.com/msalwen), if you are retrieving the access token on Azure Arc by yourself, it is recommended to use MSAL: https://msal-python.readthedocs.io/en/latest/#managed-identity

Thank you @jiasli. I have reimplemented our code following your suggestion. However, please note that there is a small but significant problem with the implementation of `msal.ManagedIdentityClient` that requires a workaround (for at least some users). In this object the method `_get_arc_endpoint` checks whether the path /var/opt/azcmagent/bin/himds exists. On our Arc servers this path does not exist. The executable it is looking for is instead in /opt/azcmagent/bin. The workaround we use is to create the directory /var/opt/azcmagent/bin and _hard_ link himds from /opt/azcmagent/bin into /var/opt/azcmagent/bin. This approach allows the code to run as intended both on the Arc server itself and within a docker container running on that server with the run flags as given in my previous comment. Of course, users should check whether the locations on their Arc servers are as we have found them or as anticipated in msal.

With this adjustment made, we can reimplement the function `_get_access_token` given above as:

```
### imports not included above that are now required ###
### Note that we no longer need to import json or os ###

import msal

from requests.adapters import HTTPAdapter, Retry

###########################################


def get_access_token(scope):
    scope = f'https://{scope.replace("https://", "")}' # allow user to pass scope with or without protocol
    mi = {"ManagedIdentityIdType": "SystemAssigned", "Id": None}
    retries = Retry(
        total=3, backoff_factor=0.1, status_forcelist=[429, 500, 501, 502, 503, 504]
    )
    sess = requests.Session()
    sess.mount("https://", HTTPAdapter(max_retries=retries))
    client = msal.ManagedIdentityClient(mi, http_client=sess)
    result = client.acquire_token_for_client(resource=scope)

    return AccessToken(result["access_token"], expires_on=int(result["refresh_in"]) + int(result["refresh_on"]))
```
which is admittedly somewhat tidier.

@richeney @husamhilal Nothing I've done here moved the needle on being able to log into azure cli using the --identity option, including setting all sorts of environment variables mentioned in various parts of the msal code and docs. The code is still looking for the identity endpoint at the IP 169. ..., which is only reachable within Azure.

**Comment by jiasli** [comment]
> will the preview of [#29187](https://github.com/Azure/azure-cli/pull/29187) enable `az login --identity`?

@richeney, yes. We plan to include it in Azure CLI 2.67.0 which will be released on 2024-11-19: https://github.com/Azure/azure-cli/milestone/148

**Comment by jiasli** [comment]
@msalwen, I saw you have created https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/762 for the incorrect `/var/opt/azcmagent/bin/himds` detection path. Once https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/763 is released, Azure CLI will pick up a new MSAL.

**Comment by husamhilal** [comment]
@jiasli  is there any (private) preview version I can test, please? (I'm internal to Microsoft)  Thank you!

**Comment by jiasli** [comment]
I have merged https://github.com/Azure/azure-cli/pull/29187 into the `dev` branch. **Managed identity support on Azure Arc-enabled Windows server** will be included in this sprint's release:

https://github.com/Azure/azure-cli/milestone/148

> Official Release: 11/19/2024
> Azure CLI version: 2.67.0

@husamhilal, thank you for your interest. You may find the edge build in the CI pipeline's artifacts:

https://dev.azure.com/azclitools/public/_build/results?buildId=203919&view=artifacts&pathAsName=false

- ZIP: https://artprodwus21.artifacts.visualstudio.com/A7b238909-6802-4b65-b90d-184bca47f458/5147fa83-336e-44ef-bbe0-c86b8ae86cbb/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2F6Y2xpdG9vbHMvcHJvamVjdElkLzUxNDdmYTgzLTMzNmUtNDRlZi1iYmUwLWM4NmI4YWU4NmNiYi9idWlsZElkLzIwMzkxOS9hcnRpZmFjdE5hbWUvemlwLXg2NA2/content?format=file&subPath=%2FMicrosoft%20Azure%20CLI.zip
- MSI x64: https://artprodwus21.artifacts.visualstudio.com/A7b238909-6802-4b65-b90d-184bca47f458/5147fa83-336e-44ef-bbe0-c86b8ae86cbb/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2F6Y2xpdG9vbHMvcHJvamVjdElkLzUxNDdmYTgzLTMzNmUtNDRlZi1iYmUwLWM4NmI4YWU4NmNiYi9idWlsZElkLzIwMzkxOS9hcnRpZmFjdE5hbWUvbXNpLXg2NA2/content?format=file&subPath=%2FMicrosoft%20Azure%20CLI.msi

**Comment by richeney** [comment]
Fantastic news. Managed identity on Azure Arc-enabled Windows will be a big win.

I look forward to Azure Arc-enabled Linux being added in once they've updated the MSAL library for Python and finishing this one off!

**Comment by husamhilal** [comment]
> I have merged [#29187](https://github.com/Azure/azure-cli/pull/29187) into the `dev` branch. **Managed identity support on Azure Arc-enabled Windows server** will be included in this sprint's release:
> 
> https://github.com/Azure/azure-cli/milestone/148
> 
> > Official Release: 11/19/2024
> > Azure CLI version: 2.67.0
> 
> [@husamhilal](https://github.com/husamhilal), thank you for your interest. You may find the edge build in the CI pipeline's artifacts:
> 
> https://dev.azure.com/azclitools/public/_build/results?buildId=203919&view=artifacts&pathAsName=false
> 
> * ZIP: https://artprodwus21.artifacts.visualstudio.com/A7b238909-6802-4b65-b90d-184bca47f458/5147fa83-336e-44ef-bbe0-c86b8ae86cbb/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2F6Y2xpdG9vbHMvcHJvamVjdElkLzUxNDdmYTgzLTMzNmUtNDRlZi1iYmUwLWM4NmI4YWU4NmNiYi9idWlsZElkLzIwMzkxOS9hcnRpZmFjdE5hbWUvemlwLXg2NA2/content?format=file&subPath=%2FMicrosoft%20Azure%20CLI.zip
> * MSI x64: https://artprodwus21.artifacts.visualstudio.com/A7b238909-6802-4b65-b90d-184bca47f458/5147fa83-336e-44ef-bbe0-c86b8ae86cbb/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2F6Y2xpdG9vbHMvcHJvamVjdElkLzUxNDdmYTgzLTMzNmUtNDRlZi1iYmUwLWM4NmI4YWU4NmNiYi9idWlsZElkLzIwMzkxOS9hcnRpZmFjdE5hbWUvbXNpLXg2NA2/content?format=file&subPath=%2FMicrosoft%20Azure%20CLI.msi

Awesome, thank you so much @jiasli , it worked perfectly! Congratulations, and well done!

**Comment by richeney** [comment]
Amazing. I look forward to testing this out!! Thanks @jiasli.

**Comment by jiasli** [comment]
https://github.com/Azure/azure-cli/pull/30267 has been merged into the `dev` branch. Managed identity support on Azure Arc-enabled Linux server will be included in 2.68.0 release (2024-12-10).

**Comment by sijjay** [comment]
> [@richeney](https://github.com/richeney), you are mostly welcome. I am also very glad this feature can be helpful to you.
> 
> If you like, you may test this functionality by installing Azure CLI from source code:
> 
> git clone https://github.com/Azure/azure-cli --depth 1
> python3 -m venv azure-cli-env
> . azure-cli-env/bin/activate
> pip install --no-deps --editable azure-cli/src/azure-cli-core azure-cli/src/azure-cli azure-cli/src/azure-cli-telemetry
> pip install --requirement azure-cli/src/azure-cli/requirements.py3.Linux.txt
> Once you have finished the testing, you may delete these folders:
> 
> rm -rf azure-cli azure-cli-env
> DEB and RPM packages are also available at https://dev.azure.com/azclitools/public/_build/results?buildId=207433&view=artifacts&pathAsName=false&type=publishedArtifacts

Awesome. Works on Ubuntu ARM64 (jammy). Thank you!
